core/src/main/java/hudson/FilePath.java

```diff
@@ -1271,7 +1271,7 @@ public void deleteRecursive() throws IOException, InterruptedException {
         private static final long serialVersionUID = 1L;
         @Override
         public Void invoke(File f, VirtualChannel channel) throws IOException {
-            deleteRecursive(deleting(f));
+            Util.deleteRecursive(fileToPath(f), path -> deleting(path.toFile()).delete());
```

Comment:
I don't think you want to call `File.delete` in the filter, since it will attempt to delete the file, and if it fails, you'll won't get an exception, since it is an old IO, not an NIO call, right? You probably want something more like: 

```
path -> {
try { 
    deleting(path.toFile())
} catch (Exception e) { // Or maybe specifically SecurityException
    return false; // Better to propagate errors though...
}
return true; // If allowed, let `Files.deleteIfExists` handle it for better error messages in case of failure.
```

That said, in the interest of propagating the actual error to users, I would suggest using a functional interface with a method like `void check(Path p) throws SecurityException` instead of `Predicate<Path>`.

Reply:
Yeah, I was considering making a functional interface like that.

And good catch. I mixed up the delete() that performs the security check with the delete() that actually deletes the file.